### ISSUE TITLE
flatten images - merge multiple layers into a single one

### ISSUE THREAD

**Issue by unclejack** [docker/docker#332]
There's no way to flatten images right now. When performing a build in multiple steps, a few images can be generated and a larger number of layers is produced. When these are pushed to the registry, a lot of data and a large number of layers have to be downloaded.

There are some cases where one starts with a base image (or another image), changes some large files in one step, changes them again in the next and deletes them in the end. This means those files would be stored in 2 separate layers and deleted by whiteout files in the final image.

These intermediary layers aren't necessarily useful to others or to the final deployment system.

Image flattening should work like this:
- the history of the build steps needs to be preserved
- the flattening can be done up to a target image (for example, up to a base image)
- the flattening should also be allowed to be done completely (as if exporting the image)

**Comment by bbinet** [comment]
+1 for AUTOCOMMIT OFF, COMMIT

**Comment by monokrome** [comment]
This still seems a bit unnatural to even support auto-committing. Doesn't it seem to be easily made backwards-compatible by defaulting auto-commits to off until the first `COMMIT` occurs.

Once the first `COMMIT` occurs, where is the value of ever needing to enable auto-committing again? This seems like a really weird edge-case that I can't really see as useful and it seems to be adding a lot more potential overheard.

**Comment by txomon** [comment]
You have a point there =) but I suppose that if you want to interpolate
auto lines with non-auto you would need something like that

On Thu, Jan 8, 2015 at 11:05 PM, Brandon R. Stoner <notifications@github.com
> wrote:

> This still seems a bit unnatural to even support auto-committing. Doesn't
> it seem to be easily made backwards-compatible by defaulting auto-commits
> to off until the first COMMIT occurs.
>
> Once the first COMMIT occurs, where is the value of ever needing to
> enable auto-committing again? This seems like a really weird edge-case that
> I can't really see as useful and it seems to be adding a lot more potential
> overheard.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/issues/332#issuecomment-69258143>.
>



-- 
Javier Domingo Cansino

**Comment by monokrome** [comment]
@txomon I think that you only want to do that for the concern of backwards compatibility, but my only concern there is that supporting it as a feature might end up being a lot more overhead and I'm not sure what the value of the feature is, and I feel like documentation may get really confusing just by trying to explain that part of it.

**Comment by txomon** [comment]
Docker creates a new layer with each Docker command. Using AUTOCOMMIT OFF
deactivates this and you will have to commit explicitly with COMMIT or
activate once more autocommit with AUTOCOMMIT ON

**Comment by monokrome** [comment]
@txomon Yeah, I understand that. I'm just wondering if anyone knows of a case where it's actually going to be useful.

**Comment by txomon** [comment]
I do, when installing packages, and need to install compile dependencies
and compile, and then you want to continue configuring and executing
commands normally

**Comment by tve** [comment]
One of the really, really painful things with docker right now is building containers where the build process needs credentials, for example to retrieve private github repos. It's difficult to avoid having the creds end up in the image. Often only three statements need to be executed with the creds, e.g. ADD the creds, RUN whatever pulls the private github repos, RM the creds. Being able to turn autocommit off around that sequence would solve the problem. I want autocommit in general, I just don't want to commit while I have the creds in the filesystem.
(I have seen other proposals to address this issue, such as nested builds and I'm fine with either approach, we just need some approach pretty badly.)

**Comment by monokrome** [comment]
@tve I've never seen that approach as a problem, but I suppose that example seems like an interesting use-case. My guess is that you need to temporarily use credentials to communicate with a remote service or something?

**Comment by tve** [comment]
@monokrome Yup. Pull from private github repos, for example. Typically done through a dependency tool, such as bundler in the ruby world, hence it's not appropriate to pull locally before kicking off the container build and then adding the pulled stuff explicitly.

**Comment by cyphar** [comment]
I reckon the solution to this problem is incredibly simple. All we need to do is create a "flatten" container from the image being squashed to, and then just apply each layer to that container and then commit it (after adding all of the metadata). Something like this:

Given a history like this (where each `L*` is a layer, `scratch` is the empty base layer and `C*` is a container) and we want to squash `L2...L5` (this syntax is inclusive):

```
scratch -> L1 -> L2 -> L3 -> L4 -> L5
```

We create a container `C1` from `L1`, then apply `{L2, L3, L4, L5}` to `C1` (both the metadata and the actual filesystem diffs). So now you have `C1` containing the changes `L2...L5`. Then we commit `C1`, creating a layer `L1A` which contains all of the changes:

```
               -> L1A (contains {L2, L3, L4, L5} and L1 is the parent)
              /
scratch -> L1 -> L2 -> L3 -> L4 -> L5
```

**Comment by cyphar** [comment]
@tve Your problem is actually going to be better solved with #10310 -- which is basically ethereal data for builds and containers that isn't saved in images.

**Comment by cyphar** [comment]
I reckon the solution to this problem is incredibly simple. All we need to do is create a "flatten" container from the image being squashed to, and then just apply each layer to that container and then commit it (after adding all of the metadata). Something like this:

Given a history like this (where each `L*` is a layer, `scratch` is the empty base layer and `C*` is a container) and we want to squash `L2...L5` (this syntax is inclusive):

```
scratch -> L1 -> L2 -> L3 -> L4 -> L5
```

We create a container `C1` from `L1`, then apply `{L2, L3, L4, L5}` to `C1` (both the metadata and the actual filesystem diffs). So now you have `C1` containing the changes `L2...L5`. Then we commit `C1`, creating a layer `L1A` which contains all of the changes:

```
               -> L1A (contains {L2, L3, L4, L5} and L1 is the parent)
              /
scratch -> L1 -> L2 -> L3 -> L4 -> L5
```

**Comment by thaJeztah** [comment]
@cyphar would that be an "external" tool, or integrated in the builder? Would signing the images still work?

**Comment by cyphar** [comment]
@thaJeztah Surely for signing images, you'd just need to sign the new squashed images?

**Comment by cyphar** [comment]
(continued because Github won't let me post the full comment ... weird)

I think we should just deal with making a command-line subcommand for doing this and deal with a `Dockerfile` syntax later. The syntax should be obvious: `docker flatten start_image...end_image` and then we just do the above magic. Obviously we'd need to do some checks to make sure that the layers on the stack are actually part of the same history, but that should be fairly trivial.

The only issue I can see is if someone wants to squash `L2...L4` with `L5` being a layer atop the squashed layers ... Maybe we could automate the generation of those layers atop the squashed layers (by just applying them on new containers and then generating images)?

**Comment by cyphar** [comment]
I reckon the solution to this problem is incredibly simple. All we need to do is create a "flatten" container from the image being squashed to, and then just apply each layer to that container and then commit it (after adding all of the metadata). Something like this:

Given a history like this (where each `L*` is a layer, `scratch` is the empty base layer and `C*` is a container) and we want to squash `L2...L5` (this syntax is inclusive):

```
scratch -> L1 -> L2 -> L3 -> L4 -> L5
```

We create a container `C1` from `L1`, then apply `{L2, L3, L4, L5}` to `C1` (both the metadata and the actual filesystem diffs). So now you have `C1` containing the changes `L2...L5`. Then we commit `C1`, creating a layer `L1A` which contains all of the changes:

```
               -> L1A (contains {L2, L3, L4, L5} and L1 is the parent)
              /
scratch -> L1 -> L2 -> L3 -> L4 -> L5
```

(continued in next comment, because >GitHub)

**Comment by cyphar** [comment]
(continued because Github won't let me post the full comment ... weird)

I think we should just deal with making a command-line subcommand for doing this and deal with a `Dockerfile` syntax later. The syntax should be obvious: `docker flatten start_image...end_image` and then we just do the above magic. Obviously we'd need to do some checks to make sure that the layers on the stack are actually part of the same history, but that should be fairly trivial.

The only issue I can see is if someone wants to squash `L2...L4` with `L5` being a layer atop the squashed layers ... Maybe we could automate the generation of those layers atop the squashed layers (by just applying them on new containers and then generating images)?

**Comment by thaJeztah** [comment]
`L5` could be an issue yes; `L5` (based on `L4`) and `L5` (based on `L1A`), basically are the same, but a different parent, which makes them...different. For *caching*, I think both need to exist, but I'm not sure.

**Comment by cyphar** [comment]
@thaJeztah Okay, I should've called the `L5` which comes from `L1A` `L5A` because it wouldn't technically be the same image (the parent data would be different).

**Comment by Maniacal** [comment]
+1 for COMMIT OFF | COMMIT ON | COMMIT
or
AUTOCOMMIT OFF | AUTOCOMMIT ON | COMMIT

It wouldn't confuse me to have COMMIT be a single keyword that does all.  But I can also see how the AUTOCOMMIT approach might make more sense.  

Also, I have a use case for turning auto-commit back on after having it off. When using large installers that you don't want to leave behind after installation but then want normal commit functionality once that step is complete.

**Comment by cyphar** [comment]
@thaJeztah Also, for caching I think the only way you could actually use the flattened images as a cache is if you `FROM` and image that is based on the flattened images...

**Comment by thaJeztah** [comment]
@cyphar yes, that would make sense. Basically that would be the same as `L5A`, which (being a layer) als can be seen as an image that depends on the flattened images.

If I understand correctly,  both `L1` to `L4` *and* `L1A` would be cached/exist *separately* after the build?

**Comment by jjn2009** [comment]
Dockerfile commands  such as ```commit (alias)``` + ```squash (alias)``` + some keywords like ```root```  ```local``` and ```end``` could be convenient. Flipping autocommit on/off is simple but a commit aliasing system might lend some more powerful features.

then you could also reference these locations from the command line (and uuids) to squash things locally before sending them off:

```docker squash root:local (image) (new-min-image)```  # squash the root image(s), everything you got using FROM
```docker squash before_secrets:after_secrets_cleaned (image) (no-secrets)``` # use your secrets, clean them up 

Potental Dockerfile:
```
FROM xyz
#add more libs
#clean to your liking
SQUASH root

COMMIT before_secrets
#secrets
#clean secrets
COMMIT after_secrets_cleaned
SQUASH before_secrets
```

It wouldn't be as easy to script a squash from bash with only uuids, if you rebuild without cache your image will likely end up with different uuids.

Theres some namespace problems with the commit names if you can reference across image but those could be easily solved by prepending some additional identifier, and making some best guess assumptions on the arguments provided in the event of a collision. 

With commit names you could also specify something like ```docker build --no-squash``` but still retain some extra metadata to use at a later date.

You could run into problems where commit names were squashed out of existence however, that could be tricky, maybe the new layer can retain the aliases in an array to prevent errors later down the line.

**Comment by monokrome** [comment]
@jjn2009 I'm a bit confused. Is there difference between the behavior `SQUASH` and `COMMIT`? Does `COMMIT` take on a new meaning where it's just setting a name here?

If so, then what is the intention of having to write the name after the `SQUASH`? This might be a bit more complicated than necessary when you can just label the `COMMIT` rules from earlier examples?

**Comment by mhennings** [comment]
Hello all,

I think there is a valid reason for differencing between squash and commit.

squash: like in git, combine layers after they were created
commit: finish this layer (unless auto commit is enabled)

So when committing you finish a layer.
With squash you can melt them together afterwards.

While it may sound theoretical, that are very distinct operations, that
could be of use in the one or other situation.

Kind regards,

Marco



2015-01-30 0:46 GMT+01:00 Brandon R. Stoner <notifications@github.com>:

> @jjn2009 <https://github.com/jjn2009> I'm a bit confused. Is there
> difference between the behavior SQUASH and COMMIT? Does COMMIT take on a
> new meaning where it's just setting a name here?
>
> If so, then what is the intention of having to write the name after the
> SQUASH? This might be a bit more complicated than necessary when you can
> just label the COMMIT rules from earlier examples?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/issues/332#issuecomment-72127858>.
>



-- 
Marco Hennings
Dipl.-Ing.

freiheit.com technologies gmbh
Budapester Straße 45
20359 Hamburg / Germany
fon   +49 40 / 890584-0
fax   +49 40 / 890584-20
Hamburg HRB 70814

0708 1C60 2620 915A 92AA 3546 ABAE 6695 7393 489E
Geschäftsführer: Claudia Dietze, Stefan Richter

**Comment by monokrome** [comment]
@mhennings I guess that this leaves me wondering where you'd want to `COMMIT` if you're going to `SQUASH` later anyway. Is there a specific use case where this is useful?

**Comment by mhennings** [comment]
@monokrome 
I think the only reason would be caching. if commits can be used for replaying changes without doing the work again, it makes sense to commit in between expensive steps.

**Comment by jjn2009** [comment]
@monokrome 
```commit``` would essentially be a meta-tag, or an alias to a commit that already happens by default. That wording was some what misleading I suppose, but I think of it like a git commit. as @mhennings pointed out. The difference between git and docker is that commits happen by default and without a message (in my proposal we make more of an easily recalled identifier not a few sentences about the change you made in git). You would explicitly be saying I'm committing this name to this point in the dockerfile, not rebasing to the last commit command in the process, so you have references to costly before or after points in the build, or points where you did you cleaning of sensitive or unnecessary data.

```squash``` would specify that you want things to be smashed together. So like git rebase, but rebase is more powerful in git than my proposal, I'm not sure if you need the full power of git rebase in docker, I wouldn't think you would want to do any sort of merging in docker, it'd be interesting to hear an argument in favor of that.

Both ```commit``` and ```squash``` could exist in the command line I suppose as well, I don't personally know the technical details of that, squash is evidently possible after the build.

This leaves some freedom to people further down the build chain, which is important when you have large teams who have differing opinions on how things should work, who might say no I don't want to squash like that so ```--no-squash``` and I'll specify the squashes with your commits and maybe add more commits to the docker file to help facilitate my needs, or team members who have differing amounts of time to commit (no pun intended) to understanding what the docker cache means for a build pipeline so a dev can just add them like comments and let someone else deal with figuring out whats optimal.

Also if I'm doing local dev. I want that whole chain of events cached, ```A -> B -> C -> D``` because I might choose to change a ```RUN``` command at ```B``` in my dockerfile quite a few times before I figure out whats optimal.

If I'm a sysadmin I'm going to want to figure out the easiest way to squash that image down so I can send it to my servers in a small package. 

The lines between dev, test, and sysadmin will continue to fade. I find myself editing docker files as a dev and certainly test/sysadmins will as well. All participants should be able to contribute meaningfully to the dockerfile beyond just a comment.

**Comment by jjn2009** [comment]
@mhennings 
I reread your comment and it seems you have something slightly different in mind but an interesting idea as well. 
I think backwards compatibility might be a slight issue however. Someones going to add their first and only ```commit``` to their dockerfile and the behavior will have to be either it smashes all the way to the beginning of the file or the root and that might cause some confusion if someone doesn't fully read the docs. That would turn off 'autocommit' before the commit command, which has been the only option.

**Comment by mhennings** [comment]
@jjn2009 
If one uses COMMIT we can conclude that he means to explicitly commit, but its probably better to have a warning or error in the case, the dockerfile contains COMMIT but not a disable autocommit.

**Comment by jjn2009** [comment]
@mhennings 
Ahh ok, I fully see your idea now.

**Comment by monokrome** [comment]
@mhennings Doesn't docker already do this well?

@jjn2009 I feel like the git analogy might be overcomplicating the feature again. Could those use-cases be served just as well with something akin `docker build --one-commit-per-run` and not make the Dockerfile format more complicated?

**Comment by mhennings** [comment]
@monokrome 
--one-layer-per-run would do it, though i think its a general improvement if we have "operations" on layers separated from the commands run inside a layer. 

My main concern is that we should not feel the need to combine multiple commands into one run, just because every RUN adds overhead. Either an option, or a specific way to commit and squash can fix this.

**Comment by blaggacao** [comment]
I think the underlying metastructure in our minds which drives this generic feature request might look something arround this (not well tought through):

- PREPARE: user, permissions, repository, ssh-keys
- BASE TOOLING: (curl, ssh, etc) - not needed by the app
- BUILD TOOLING: temporarily needed stuff to pre-/compile the app
- RUNTIME: apt-get and pip which is needed during app runtime
- STACK: additional stack tooling, like wsgi, but on a single instance also - exceptionally - : nginx, psql
- APP SOURCES: getting tagged sources from git directly
- CONFIG PROPAGATION: use templating, unless you have a K/V native configuration mechanism
- CLEANUP: remove all build tooling and strip off unnecessary distro-defaults to reduce attac vector and image size.
- EXPOSURE: Volumes, Entrypoint, CMD, Expose, but also define defaults for interfaceing ENV variables for transparency, if there are many, an env-file could be provided instead

While discussion functionality here, the motivation would be even better solved when combining functionality with default: batteries included but removable, this could mean a set of very generic SECTION directives which could be coupled to squashes (to make the technical link to this discussion). Build in best preactices, though probably opnionated, are just tremendously effective when it comes to (implicit) knowledge transmission. Removable, so that no one has to vory the "opinionated" notion.

**Comment by ajodock** [comment]
How about using braces?

```
{
  ADD file.tar.gz /root/
  WORKDIR /root/file
  RUN make
  RUN make install
  RUN rm -rf /root/file
}
```

So any action in the braces is run in one commit.  As much as I would like this syntax more the main disadvantage of this vs using keywords is that keywords can be run through older versions of docker and it will just skip over them, where as braces will break the Dockerfile.

In general being able to choose when we commit would also mean that we would no longer need to follow this from the Dockerfile best practices page:
>  Because image size matters, using ADD to fetch packages from remote URLs is strongly discouraged; you should use curl or wget instead.

I think that recommendation (which makes sense right now) causes a lot of problems.  i.e. say I am behind a http proxy.  The container doesn't know anything about the proxy so the curl fails with a connection refused.  I would rather not modify the Dockerfile to work in just my environment (and possibly have to code username/passwords into your containers), because that defeats the purpose of Docker!  

With any multiple commands per commit we can switch best practices to using ADD/COPY instead because the Docker service does the downloading which you presumably have configured for your proxy for grabbing images.

It would also save time by not needing to install tar+gz in every container just to remove it again in the same step since Docker will handle the untar with the ADD command instead.

**Comment by gdm85** [comment]
@blaggacao that's a nice summary of common image sections, thanks for  sharing.

@ajodock I immediately thought that ENV should be local to those curly braces. There's to say that the idea of curly braces might not get traction as it would add "depth" to Dockerfiles, but I still like it.

**Comment by jdmarshall** [comment]
Adding logical blocks to Docker without adding any indentation or block separator would be pretty regressive.  We tried that in the 60's and by the 70's virtually every computer language or scripting language had realized how bad an idea that was.  

MIME is the only big exception I can think of, but that's pretty much only ever read by machines.

**Comment by monokrome** [comment]
#NoDepthForDockerfiles

**Comment by monokrome** [comment]
#NoDepthForDockerfiles

**Comment by muayyad-alsadi** [comment]
I plus COMMIT OFF | COMMIT ON | COMMIT

what is the status of this long thread

**Comment by monokrome** [comment]
I still don't see a practical usecase to ever need `COMMIT OFF`, which kind of leaves me feeling like `COMMIT ON` could be implicit with just a `COMMIT`.

**Comment by txomon** [comment]
Installation through compilation of one application would be one of the
first steps, but the configuration might be better to be apart.

On Mon, Apr 27, 2015 at 9:10 PM, Brandon R. Stoner <notifications@github.com
> wrote:

> I still don't see a practical usecase to ever need COMMIT OFF, which kind
> of leaves me feeling like COMMIT ON could be implicit with just a COMMIT.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/issues/332#issuecomment-96783006>.
>



-- 
Javier Domingo Cansino

**Comment by cyphar** [comment]
I don't feel that blocks should be represented using start and end keywords. This isn't the 80s. Just make it a simple `{ ... }` block.

**Comment by thaJeztah** [comment]
> I don't feel that blocks should be represented using start and end keywords. This isn't the 80s. Just make it a simple { ... } block.

In *general* I agree. Haven't really made my mind up for Dockerfiles; Using `{..}` looks a bit out of place (visually) with the overall syntax of the Dockerfile, which *does* use keywords for everything.

Note that there is currently an open PR/Proposal for `MARK ... SQUASH` here: https://github.com/docker/docker/pull/12198 and allows "squashing" layers.

**Comment by jdmarshall** [comment]
I originally voted for making either the start or end be implicit so that only one marker was needed.  Now I'm wondering if you shouldn't just have a flag that squashes the whole thing.

If I want 2 layers, I would make 2 separate docker configurations, one of which is FROM the other one.

**Comment by monokrome** [comment]
I don't think it's a "block" at all. Having a start and end keyword makes it feel like a block, but it doesn't need to feel like one. I feel like it really only needs on keyword that denotes when to separate things. It seems to me that having a dozen Dockerfiles seems like less maintainable then just dealing with all the unnecessary layers, so it kind of brings us back to the original issue here.

**Comment by borromeotlhs** [comment]
How about multiple from statements where From calls allow for pruned opaque
images?

allows for existing syntax and doesn't introduce having to clean up RUNs in
dockerfiles, but rather, on images themselves where it seems more
appropriate as you already intend those things to be opaque to end users?

**Comment by monokrome** [comment]
@borromeotlhs I think that your suggestion may fit a bit into [this issue][incld], but coupling the idea of inclusion and per-file squashing seems dangerous to me. The explicitness of a Dockerfile like this seems like it might be better for users:

```
FROM ubuntu
MAINTAINER Brandon R. Stoner <monokrome@monokro.me>

INCLUDE Dockerfile.dependencies
COMMIT

INCLUDE Dockerfile.base
COMMIT

INCLUDE Dockerfile.web
COMMIT
```


incld: https://github.com/docker/docker/pull/12749 "Support INCLUDE in Dockerfiles"

**Comment by borromeotlhs** [comment]
@monokrome
I actually like the idea of making sub docker files with a naming scheme of
Dockerfile.{whatever} with the possibility that that file could be an idiom
laced text based file (for common reuse) or a opaque binary object.  In
either case, I feel COMMIT syntax would be redundant.

Also, if pruning a docker image or dockerfile is dangerous, you'll have to
give more info o. why, as I'm not seeing how it is more dangerous than
leaving sensitive data _in_?

**Comment by monokrome** [comment]
@borromeotlhs I think that the `COMMIT` syntax is only redundant in the previous example, but without it it the possibility of the following is impossible:

```
INCLUDE Dockerfile.dependencies
INCLUDE Dockerfile.base
COMMIT

INCLUDE Dockerfile.web
COMMIT
```

The last paragraph in my previous comment explains the pitfalls of assuming a `COMMIT`. I don't feel like `INCLUDE` being coupled to ` COMMIT` syntax would make much sense at all. An `INCLUDE` is just directly injecting the contents of another file - which may also have it's own `COMMIT` statements.

**Comment by tcurdt** [comment]
Please - no separate files (if required as such).

    ADD file.tar.gz /root/
    WORKDIR /root/file
    RUN make
    SQUASH
    RUN make install
    RUN rm -rf /root/file
    SQUASH

Why not just `SQUASH` (or `COMMIT`)

**Comment by borromeotlhs** [comment]
Squash would flatten images, but my original goal with flattening was to
hide some necessary steps to create a layer that I don't want others to
peer into.  hence, the whole inclusion of another dockerfile that, say,
injects some public and private key info into containers so they are
automatically able to access another external site (s3,etc) is good as I
can then dictate, via multiple FROMS to incorporate another Dockerfile that
I as an end user may not have control over but could pull like images.

Commit on/off will definitely flatten images, though the need to constrain
layers to fit under arbitrary limited seems to be going away.

As of right now, we also "assume commit" regardless, so I again don't see
how it's bad or dangerous.

Being able to add opacity akin to the from line seems to be a win as it
also doesn't force brackets, on/off keywords or delimiters of any sort.

Cheers,
TJ

**Comment by jdmarshall** [comment]
@monokrome  dozen?  Kind of hyperbolic don't you think? 

I can't imagine anyone ever using squash and still ending up with twelve layers at the end of the process.  I was thinking three or four (before, secret, after). What sort of docker image are you making?

**Comment by monokrome** [comment]
@jdmarshall I think that it's not dramatic to suggest that a dozen Dockerfiles would be used in a complicated production system isn't outlandish. With that said, I also think that the only rhetoric here is that of the person pointing out a single word in a previous comment instead of trying to understand the suggestion which I was trying to make.

I don't see 4 Dockerfiles solving the problem any more elegantly, and I feel like any number greater than 1 potentially exhibits the same result. Possibly even 1 does.

**Comment by jdmarshall** [comment]
@monokrome

| It seems to me that having a dozen Dockerfiles seems like less maintainable then just dealing with all the unnecessary layers, so it kind of brings us back to the original issue here.

Of course later you talked about another way to have multiple files so now I don't know what to think.

A dozen is a lot. In most software that is code for "too many" and also seems to be the argument you're making, so I'm not hanging on a word, I'm hanging on an inequality and your unstated but very clearly implied "no". 

Many of us are only concerned about a couple of layers.  For some that's apt-get churn, for others it's an authenticated GET request, or a password.  "A couple" is not just doable, it's reasonable.  

Far more reasonable than introducing keywords that affect the semantics of the statements between them (known generally as Blocks) and not calling a spade a spade.  

There are no block semantics in Docker now, aside from base images.  Please think twice before setting a precedent that will encourage people to ask for more block semantics.  

But if you still must add them in the end, don't half-ass it.  Call a spade a spade.

Also, if you are honestly concerned about too few layers slowing transfer time, I wouldn't worry too much.  First we're kinda here talking about trying to make smaller images, so it should be a net gain.  But if push comes to shove we could probably figure out how to do simultaneous Range requests (like the DownloadThemAll browser addon does), but if everyone is behaving the same way it's all a wash.

Me, I have a crappy network connection at home and often so do the people I'm trying to pitch Docker to, as part of their toolchain. So I care more about total size than throughput.  I also care about credentials but I work around that (would be nice to get rid of the workarounds though)

**Comment by jdmarshall** [comment]
@borromeotlhs if you want to run your sensitive code at the beginning of someone else's docker build that might work, but if you run it in the middle of theirs you can't protect your credentials.

I could just swap out curl and sftp and ssh and git on my image and copy your keys without you being able to detect it.  Stay with the base image idea if you want to avoid trouble.

**Comment by borromeotlhs** [comment]
@jdmarshall, I disagree that more docker files somehow equates to more
total bandwidth.  if anything, you end up using less bandwidth, especially
if those dozen or so dockerfiles just to common idioms that are necessarily
private.

I agree that I don't like block syntax though, and it is under this precept
that I offer up the multiple FROM idiom.  it is succinct, there are no new
keywords, it works at first glance how one would suspect from visual
inspection of a parent dockerfile, it is reusable, and I feel that it could
be pulled (and stored locally like images === no extra bandwidth ) into
whatever containers needed it.

Now, I'm again only interested in deleting layers due to the presence of
things that I don't want in a shipping dockerfile.  I don't necessarily
care about minimizing layers as I think later limitations are a problem
that will solve itself in the future.

**Comment by jdmarshall** [comment]
@borromeotlhs re: lower bandwidth, I agree entirely. Sorry if that wasn't clear.

**Comment by borromeotlhs** [comment]
no prob.  I just love the serious consideration this feature is getting :)

**Comment by jdmarshall** [comment]
@monokrome I think essentially I'm saying that I've come around to the --one-layer-per-run solution brought up back at the end of January (I forgot that had been put forward). Are you no longer in favor with this solution?

**Comment by jdmarshall** [comment]
@blaggacao +1 on the list either flattening the whole thing or making the entire lifecycle explicit would solve a bunch of problems.  I didn't assume we could know a priori what those stages would be but maybe we actually can...

I'm not at all happy with the heavy handed usage of && in RUN.  Everybody does it and everyone who writes about docker tells more people to do it as well.  Chaining has all of the problems of squashed layers and is harder to read to boot.

**Comment by txomon** [comment]
Just remember that if I have to compile from source various libs and the
program, i may now want to have there at the end nothing bur binaries,
deleting private keys i might have used, sources etc.

You can check for use cases in previous posts, as i think we are failing in
the same arguments once and another.

**Comment by gdm85** [comment]
I've been tinkering with Docker for ~1 year now, and at some point I have removed [this image layers limit](https://github.com/docker/docker/blob/62a85fe2026f3f5114f19638baf53d19a51d39c0/image/image.go#L20) (as I am not using AUFS either) because yes, to aim to production you need a lot of steps in your images and packing instructions with concatenation '&&' or scripts is not elegant, nor readable (and against providing all the instructions in a Dockerfile, for that matter).

If this feature gets in (as SQUASH/COMMIT, with or without the block initiator/braces), I'll probably use it at the end of **each** Dockerfile that leads to a production image, while never do it for a high-flux/development image (as you want to benefit maximally from caching).

Although I'd keep any day the [removal of the layers limit](https://github.com/gdm85/docker/commit/7d1d482807d1a5232c364fb7db8466c566ba55a0), I would start using SQUASH/COMMIT because that puts back me (the user) in control of how I want to use/misuse layers.

**Comment by muayyad-alsadi** [comment]
I don't like blocks { }

dockerfiles are sequence of commands so AUTOCOMMIT ON/OFF is not a block because it's meaning less to nest them.

the use case would be like this:

```
AUTOCOMMIT OFF
ADD PRIVATEKEY
RUN git clone ...
RUN rm PRIVATEKEY
RUN rm -rf .git
RUN make
RUN find . -name '*.o'
AUTOCOMMIT ON
```

1. the objective is to not include the private key used to clone the code (we don't want to ship a layer with sensitive data)
2. we need to save space by omitting the layer with big `.git` directory and building intermediate files (we don't want to ship layers with useless files: .o files, .git directory, yum/apt cache ...etc)

I would think of layers to be

1. base distro layer
2. base dependencies
3. base app with config...etc.
4. the latest milestone release of the app
5. the latest release of the app

but I don't think they should be

1. base distro
2. distro with some changed config file (ex. repo)
3. package installed
4. package configured (ex. put `/etc/httpd/conf/httpd.conf`)
5. another packages installed
6. another package configured
7. touch a log file
8. chmod a log file
9. add user
10. add key
11. compile object files
12. link object files
....

the layers are just bunch of command results of no high level meaning

**Comment by duglin** [comment]
@gdm85 your comment got me wondering about something.  If you plan on squashing production-target images, but not dev-target images how do you see controlling that?  Simply adding a COMMIT to your Dockerfile after you're done developing it? Or do you see the same Dockerfile being used in both environments but have the need for some kind of if-statement around the COMMIT to enable/disable it based on what you're doing at that moment?

**Comment by thaJeztah** [comment]
> while never do it for a high-flux/development image (as you want to benefit maximally from caching).

@gdm85 the `MARK..SQUASH` instructions still make optimum usage of the cache; see https://github.com/docker/docker/pull/12198#issuecomment-91061494. Both individual layers *and* the squashed result are cached, so that's a win-win.

**Comment by gdm85** [comment]
@thaJeztah nice one! Thanks for pointing it out. Yes, that would be best because one could tinker around on the registry/development server, and then staging/production peers get the squashed images :)

@duglin right now I am splitting images (and common ancestors) by using different Dockerfile's, trying to organize them hierarchically as much as it is reasonable to do. I'd say yes, I would like a conditional usage - depending on whether I want to **consume** or develop on the image, so what @thaJeztah mentioned seems like a good feature for such use-case.

So - even if technically the fully-squashed image is an actual duplicate of the image made of N combined layers, I think it would be best suited for the consumption of the images (in staging/production) - also for the secrets/privacy management concerns that were mentioned by others here.

**Comment by borromeotlhs** [comment]
why even bother making commit on/off syntax visible in production
dockerfiles?  I mean, layers in images aren't even really related to a
dockerfile as one approach would be to squash the images layers apart from
messing with a dockerfile.

so, the call to take control of commit syntax in development, and then only
deal with squash or merge on production deployment points to the fact that
we _should_ have control over layers, but that end users of our dockerfile
can't get a dockerfile at all if there are any steps that contain private
information (layers which contain cruft aren't really an issue, they just
add up overall size).

So again,I think that going with multiple Dockerfiles, and treating sub
Dockerfiles as potentially binary like images, is still the right way to
go.  akin to distributing source (base Dockerfile) and then linking them to
objects when compiling ( build command with sub Dockerfile.whatever files )
would solve all use cases.

**Comment by jdmarshall** [comment]
@borromeotlhs are you tracking https://github.com/docker/docker/pull/9176 ?  

If they can sort out the cache determination I think it fulfills your concerns.

**Comment by monokrome** [comment]
@muayyad-alsadi Wouldn't this fulfill your request more elegantly?

```Dockerfile
ADD PRIVATEKEY
RUN git clone ...
RUN rm PRIVATEKEY
RUN rm -rf .git
RUN make
RUN find . -name '*.o'
COMMIT # Nothing commits until the first COMMIT when a COMMIT is present
```

**Comment by muayyad-alsadi** [comment]
Yes it does fulfill. Is it more elegant? I'm not sure because this way the
meaning of "commit" is overloaded with an implied condition. And explicit
is better than implicit. If some sort of include feature got introduced I
won't be able to see if there is commit in them or not.

**Comment by sleaze** [comment]
It's been over 2 years since this functionality was filed and there is still no working solution for this merged.. yikes!

Guess I'll just let my images grow unbounded.

**Comment by PierreR** [comment]
:+1:

**Comment by monokrome** [comment]
@muayyad-alsadi What condition is the meaning overloaded with? I think that your conception that there are actually 2 conditions is a misconception. Either you tell it when to make a commit *explicitly*, or else it would assume a commit for every action (for backwards compatibility reasons). Where is anything overloaded?

**Comment by morallo** [comment]
+1 for the COMMIT / AUTOCOMMIT ON / AUTOCOMMIT OFF proposal from @txomon.
It feels like the simplest solution, a lot more explicit than disabling AUTOCOMMIT with a COMMIT.

If I understand @monokrome alternative correctly, if you want to disable autocommit only for a specific set of instructions (adding private keys and removing them, or installing build tools and removing them), you would have to use COMMIT explicitly for every other instruction from the beginning of the file.

**Comment by zoechi** [comment]
+1 MARK..SQUASH

I like 
- `AUTOCOMMIT OFF`/`COMMIT`/`AUTOCOMMIT ON` for it's simplicity
- `MARK`..`SQUASH` for preserving cache usage

So in overall I think `MARK`..`SQUASH` is the better solution because it has the benefits of both, but might be more costly to implement.

I really hope to soon get rid of these cumbersome and error-prone
```
RUN \ 
  abc && \ 
  def && \ 
  ghi && \ 
  jkl && \ 
  mno && \ 
  pqr && \ 
  stu 
```

**Comment by ewindisch** [comment]
Rather than 'MARK..SQUASH', you could just do 'SQUASH $N' to squash $N previous layers. This would be easier to parse and would continue to treat the Dockerfile as a RPN / layered format rather than introducing branches.

**Comment by tiborvass** [comment]
* Review with @duglin @crosbymichael @diogomonica @LK4D4 *

The problem with this issue is that it provides a solution to a problem that yet has to be defined

The bigger question is what is the end goal for users who need to squash layers? Faster pull/push? Removing some layers (with secrets)? Maximum number of layers?

It feels like all these problems are implementation details and solving them by exposing squashing is probably the wrong way.

We suggest to address each independent issue you guys have, in more focused separate issues.

For relieving the immediate pain, we also suggest updating the documentation on how to manually squash layers if people really need it, with external tools. @crosbymichael volunteered on writing such a tool.

We're closing this issue. Would love to continue the debate on more focused issues.

**Comment by sleaze** [comment]
@ewindisch how am I supposed to know how many layers have been created?  Lol..just imagine if SQL forced you to recall how many statements had been executed.. `COMMIT 5` would commit the past 5 statements.

What a terrible idea, mate.

**Comment by sleaze** [comment]
@tiborvass Why should we have to continue to debate with you?  Why hasn't this been implemented?  What have you guys been doing for the past 2 years?

**Comment by stevenschlansker** [comment]
I also have been frustrated at the tendency of the Docker project to close entirely legitimate issues such as this one with non-fixes or "open a different bug later and let's start the whole process over"...

Any future discussion will either be referring to this ticket or rehashing the exact same ideas again.  Closing this ticket with "we'll split this up into smaller tickets" would be a palatable resolution if anyone had actually created said tickets, or if there was any indication that anyone actually cared to reach resolution for the numerous problems faced by people trying to run Docker, but alas we continue to despair...

**Comment by thaJeztah** [comment]
@sleaze please watch your language, and don't turn this in a "personal" issue.

We don't have to all agree on decisions that are made (I would have liked this feature myself), and you are absolutely free to express your opinion, but please do so in a reasonable way.

**Comment by thaJeztah** [comment]
> Closing this ticket with "we'll split this up into smaller tickets" would be a palatable resolution if anyone had actually created said tickets,

Agreed. Let's do so. Can you create a bullet list of topics you think make sense as separate issues here first? We can discuss them shortly and then create those, and link them from here.

I have to go in a few minutes, but will watch this issue, if needed, I can be on IRC tomorrow

**Comment by ewindisch** [comment]
@sleaze A Dockerfile is not SQL. We know what the previous statements are... they would be directly preceding the SQUASH command in the Dockerfile. I agree it's somewhat **ugly** looking and not as user-friendly.

The main disadvantage on technical merit would be that we wouldn't know how many layers were created from ONBUILD instructions.

**Comment by ewindisch** [comment]
@stevenschlansker I think that's a fair criticism. This issue was closed because it was seen as a proposed solution, where we might instead like to see the *problem* proposed and discussed. If/once squashing is decided as a solution, we might discuss the details of implementation. Reading the original report, the use-case for this particular issue was:

```
@unclejack:
There are some cases where one starts with a base image (or another image), changes some large files in one step, changes them again in the next and deletes them in the end. This means those files would be stored in 2 separate layers and deleted by whiteout files in the final image.

These intermediary layers aren't necessarily useful to others or to the final deployment system.
```

There are a number of ways to solve this problem which are NOT squashing. Alternative image formats, for instance may solve this natively, of which I have several ideas (and we've even seen one bittorrent-based PoC offer one potential solution).

I hope this helps.

**Comment by tcurdt** [comment]
To break it down I think the people discussing here want to to get rid of running everything in one single RUN just to avoid inflating the size of the image by intermediate layers and/or try to deal with the maximum number of layers limitation. Is that a fair description of the problem?

Frankly speaking I don't see how a different image format or bittorrent(??) could help with the size inflation - either you store intermediate steps or you don't.

In fact if you compare the current model of operation with a git workflow - usually you would not want a commit after each and every modification either. You group modifications in commits. Same would be nice for layers. This isn't really squashing - but controlling when layers are being created.

And I agree with @stevenschlansker - closing this issue without opening new ones that actually address the way forward was not very helpful.

**Comment by TravisCardwell** [comment]
@tiborvass @crosbymichael 

My goals in managing layers are as follows:

*Improved Dockerfiles*

In order to minimize layers, it is common practice to combine commands using
backslashes, which significantly reduces the utility of the file format.  It
is error-prone: if you insert a line and forget to add the backslash, the
build will break.  It promotes poor organization: separation of unrelated
commands results in additional layers, so they are often thrown together.  It
promotes poor documentation: adding comments in between continued lines is not
possible.

With better layer management, we would be able to write better Dockerfiles.

*Organization*

I want every layer that I create to exist for a reason.  Every layer should be
one that is `run` and/or a parent for other layers.  As a general rule of
thumb, I think that a layer probably shouldn't exist if there is no need to
tag it.

*COW-Filesystem Layer Minimization*

Increased layers increases the depth of the copy-on-write filesystem mount.  I
have never had a problem due to this, but having many unnecessary layers (such
as layers that contain only metadata that is later overridden) just seems
wasteful and avoidable.

---

Personally, I am not worried about the following:

*Faster Push/Pull*

I do not think that "squashing" layers would significantly effect transmission
speed, at least for my use case.

*Removing Layers With Secrets*

I do not use layers to maintain a history of how they were built, as I am of
the opinion that using Docker like Git is an anti-pattern.  (The deprecation
of `docker images --tree` supports this feeling.)  My images are built from
revision-controlled source files.

*Maximum Number Of Layers*

I currently avoid this issue by combining commands with backslashes.

---

For me, an optimal solution would be a flag for `docker build` that causes it
to process a Dockerfile and build a single layer instead of a layer per
command.  A separate "squash" tool would be sufficient, however, as long as it
can squash a specified range of layers (keeping a specific parent layer); I
would simply make a `docker-build` script that squashes every build that I
perform.

**Comment by borromeotlhs** [comment]
the reduction of writes to disk for unnecessary layers is a huge concern to
reduce io and general wear on hard disks.  less layers needlessly pulled
decreases time to deploy and should manner to everyone who is charged to
host their cloud runtime.

So many reasons to implement this other than getting rid of the irritating
(RUN * &&)* syntax.

**Comment by thaJeztah** [comment]
External tooling also cannot be used for automated builds on the Docker Hub Registry (https://github.com/docker/docker/pull/12198#issuecomment-102258400)

**Comment by txomon** [comment]
This issues have been already explained multiple times and yet I see this is the 3rd or 4th iteration in the loop:
1.- Explain problems
2.- Discuss a solution
3.- Bikeshed on the exact form of the solution
4.- Say the issue is not relevant because of not defined problem
5.- Start over.

I have seen there is no intention in 2 years this issue has been opened to implement it, so I will just unsubscribe, please don't @mention me, as I want to avoid getting notifications.

**Comment by foxx** [comment]
Shame this isn't supported

**Comment by monokrome** [comment]
I think that txomon's (not mentioning by request) example of the process happening here may be a bit wrong, but it's still a valid concern that this hasn't gone anywhere yet.

There are a lot of issues where this is referenced and then they are closed, but now we are all hanging out in the kitchen sink talking about 20 different things. It may be worth considering leaving the other issues open - at least until this is figured out.

It is completely possible that those issues aren't going to all be solved by the result of this thread (if it goes anywhere) or that someone may have a pull request related to another issue which maybe doesn't solve this case but solves one of the other ones. I think that it's possible that having so many different problems being solved in a single thread does not help solve them.

**Comment by zerthimon** [comment]
I'd like to have a way to unite multiple FS layers into a single layer.

**Comment by foxx** [comment]
Get in line, it ain't gonna happen any time soon, at least not properly. There are third party tools which claim to support this, but they are flakey at best.  

For now you can use something like the following, it's pretty nasty but it's the closest you'll get for now.

```bash
#!/bin/bash
echo "Flattening base image.."
SRC=myrepo:version
DST=myrepo:version_flat
ID=$(docker run -d ${SRC} /bin/bash)
docker export $ID | docker import - ${DST}
```

**Comment by zerthimon** [comment]
Thanks @foxx

**Comment by monokrome** [comment]
Well, this is an extremely disappointing way to leave dozens of GitHub issues unresolved without any suggestion of following up in the future. :-1:

**Comment by thaJeztah** [comment]
Flattening images / nested builds is still on the radar. Making changes in that area is on hold until the builder is split from the daemon; https://github.com/docker/docker/blob/master/ROADMAP.md#22-dockerfile-syntax, which is actively worked on, but requires a lot of refactoring.

**Comment by TomasTomecek** [comment]
@tiborvass wrote:
> For relieving the immediate pain, we also suggest updating the documentation on how to manually squash layers if people really need it, with external tools. @crosbymichael volunteered on writing such a tool.

Did @crosbymichael write such tool? Will it be even possible to write such tool after 1.10 is out?

**Comment by bmeng** [comment]
@foxx cool!

**Comment by foxx** [comment]
@TomasTomecek The other third party tool was [docker-squash][1], but it's /very/ unstable. The best you can hope for right now is the approach I mentioned earlier.

Sadly the `docker export` command does not support ranged extraction either, therefore you can only merge from start to X, rather than X to X. This means that if you perform a merge on every build release, you'll have to re-upload the entire image, rather than a small portion of it.

Another option is to use Docker in combination with pip, where you package up your application as a `pypi` package and then only push new Docker containers when you need to update your system libs/deps (which ideally should be at least once a day, to ensure you're getting security patches). However, this means having a bootstrapper inside your container which is then running on your production boxes, a particularly nasty devops pattern.

I'll be touching on this topic more in my next blog post.

tl;dr - It's unlikely that you'll see any improvement in this area for at least another 18 months, by which time [Rocket][2] should have reached production maturity.

[1]: https://github.com/jwilder/docker-squash
[2]: https://github.com/coreos/rkt

**Comment by TomasTomecek** [comment]
> @TomasTomecek The other third party tool was docker-squash, but it's /very/ unstable. The best you can hope for right now is the approach I mentioned earlier.

The unstability of `docker-squash` was the reason we wrote our own tool: https://github.com/goldmann/docker-scripts#squashing and we are using it in production now.

> Sadly the docker export command does not support ranged extraction either, therefore you can only merge from start to X, rather than X to X. This means that if you perform a merge on every build release, you'll have to re-upload the entire image, rather than a small portion of it.

This is the exact reason we discarded the solution.

**Comment by foxx** [comment]
Interesting, your library seems to have a decent amount of tests as well. I'll give this a mention in the upcoming article, as it looks to serve as a decent workaround.

**Comment by beorn** [comment]
I think flattening of layers is something that should be a standard part of Docker, via syntax in the Dockerfile, so that all of the vendor-provided docker images on Docker Hub can use it.  It's important to provide small base images, and it'd be a shame if that had to be done through external tools importing/exporting, losing the transparency we have in Dockerfiles describe how images came to being.

Personally, I think a simple syntax that should work is just to allow any Dockerfile command (or any where it makes sense) to be preceded with an `AND` keyword, a slightly generalized version of the [above feature request](https://github.com/docker/docker/issues/19597), e.g.,

```Dockerfile
WORKDIR /app
AND COPY requirements.txt /app/
AND RUN apt-get update
AND RUN apt-get install some-build-dependencies
# Install app
# (comments are ignored, so following commands are still in the same layer)
AND RUN pip install -f requirements.txt
AND RUN apt-get purge --auto-remove -y some-build-dependencies
AND RUN apt-get autoremove && apt-get clean

# following command not part of above layer
RUN apt-get install
```

It's easy enough to use # comments to document each layer and what's going on - that's what comments are for anyways.

**Comment by cgrandsjo** [comment]
My opinion is that when working with a Dockerfile the default behaviour should be to just create one additional layer on top of the base image by default, independent of how many commands the Dockerfile contains.

If you for some reason want to add an extra layer then there should be an ADDLAYER command available that you could insert at any line in the Dockerfile to separate the layers.


`FROM ubuntu
MAINTAINER Me <me@host.com>
INCLUDE Dockerfile.dependencies
INCLUDE Dockerfile.base
INCLUDE Dockerfile.web
`
creates one layer on top of the ubuntu image.

`FROM ubuntu
MAINTAINER Me <me@host.com>
INCLUDE Dockerfile.dependencies
ADDLAYER
INCLUDE Dockerfile.base
INCLUDE Dockerfile.web
`

Creates two layers on top of the ubuntu image.

**Comment by justincampbell** [comment]
@cgrandsjo That would cause the cache to not be used by default.

**Comment by cgrandsjo** [comment]
@justincampbell: Sorry, please elaborate your answer. Omitting the ADDLAYER command actually means that ADDLAYER is added to the end of the file and next time you build with the Dockerfile the cache will be used because a layer was created.

**Comment by mishunika** [comment]
Yay, now I have images of around 30Gigs in size, and their actual size should not be more than 10G!

So yeah, just stepped into the same difficulty, and I was thinking that Dockerfile is really lacking some kind of COMMIT action (Inspired from the DB transactions) to decide where a layer should end. 
Then I found this issue and I've read other commit related ideas that are in fact the same. I think that user should be able to specify explicitly what the layers should contain and when they should start/end.

Furthermore, in my opinion, the issue with caching is not a big one though. It can be the same as now, but delimited by the commit/addlayer levels, if something has changed in a such block, then no cache is used at all for this level. And even more, the commit thing can be optional, and the default behavior can be maintained as it is now.

**Comment by foxx** [comment]
@mishunika See my previous answer, and also @TomasTomecek, for a workaround. Don't bother trying to push this proposal with Docker, it ain't going to happen any time soon (see previous comments from core devs)

**Comment by campbel** [comment]
@TomasTomecek is https://github.com/goldmann/docker-scripts#squashing a suitable tool for reducing image size? 

For instance given a docker file:

```
FROM baseimage

RUN apt-get buildtools
ADD / /src
RUN  buildtools build /src

RUN rm -rf /src
RUN rm -rf /buildtools`
```

After building and squashing, would the resulting image lose the size of the src and buildtools?

**Comment by goldmann** [comment]
@campbel That's correct. This tool will remove unnecesary files. I haven't tested it with ADDing root filesystem (/) to the image (it's generally, a very bad idea), but I understand that this is just an example. 

Please note that Docker 1.10 is still in works (see v2 branch). Feel free to open any issues.

**Comment by yoshiwaan** [comment]
I think the AND and ADDLAYER options mentioned above are useful for certain situations (such as controlling what to and what not to cache), but if you are chaining builds from images you control and later builds are removing things from the upstream builds then they don't help with the size problem.

Something as simple as a `--squash` option to `docker build` which looks through the layers and removes whiteout files and all underlying files in above layers (correct me if I'm wrong but that's my understanding of how it works) would be extremely useful.

It's the same as when you use git really, sometimes you want to rebase, sometimes you want full commit history and sometimes you just want to squash all that noise out of there.

**Comment by sivang** [comment]
So, is this going to be a feature in docker or already solved in the stable release somehow?

**Comment by cpuguy83** [comment]
@sivang Maybe, now that the image format has been changed: #22641.

Please don't spam the PR, though.

**Comment by foxx** [comment]
@sivang I'll be surprised if you see this feature in a release before 2017. If you need a quick fix, read previous [suggestions][https://github.com/docker/docker/issues/332#issuecomment-185133695] or check out the far more superior option, [rkt](https://github.com/coreos/rkt)

**Comment by thaJeztah** [comment]
Thanks for the commercial break, @foxx

**Comment by sivang** [comment]
Well, I used jwilder's docker-squash, it seemed to have done the flatten
job but loading the image back doesn't show it on the docker images list...

**Comment by JonathonReinhart** [comment]
Yet another disappointment from the Docker team; not because some feature doesn't exist, but because of a dismissive attitude by the maintainers.

@tiborvass said (https://github.com/docker/docker/issues/332#issuecomment-102133413):

> The problem with this issue is that it provides a solution to a problem that yet has to be defined
> ...
> We're closing this issue. Would love to continue the debate on more focused issues.

Perhaps he didn't read the original issue (which was opened over three years ago), which very clearly stated:

> There are some cases where one starts with a base image (or another image), changes some large files in one step, changes them again in the next and deletes them in the end. This means those files would be stored in 2 separate layers and deleted by whiteout files in the final image.

> These intermediary layers aren't necessarily useful to others or to the final deployment system.

I don't understand what is "yet to be defined" or not focused about that, but in case you need something concrete:

```
FROM debian

# This line produces an intermediate layer 400 MB in size
ADD local_400MB_tarball_im_about_to_install.tar /tmp

# This line installs some software, and removes the tarball.
# Lets say it produces a layer with 20 MB of binaries
RUN cd /tmp && tar xf local_400MB_tarball_im_about_to_install.tar && cd foo && make install && cd /tmp && rm local_400MB_tarball_im_about_to_install.tar
```

The end result is that this image is `sizeof(debian) + 420MB` in size, when 400 MB of it were removed.

Perhaps if issues were addressed instead of dismissed, this project would have nearly *as many issues in its history as it does commits*.

**Comment by cpuguy83** [comment]
@JonathonReinhart This problem is this issue is discussing a particular solution rather than the problems.
In reality, squashing is a stop-gap to a particular problem that is an implementation detail of the current storage subsystem... ie, we don't need squashing if/when the storage subsystem is replaced with a better solution.

Thank you for your kind and thoughtful comments.

**Comment by JonathonReinhart** [comment]
@cpuguy83 Sarcasm isn't necessary when someone is expressing frustration.

Regardless of whether or not this is the right solution, people will find this issue when looking for a solution to a very common problem. When you see that the issue is **closed**, you'll immediately wonder "Why was this closed? Was it fixed?", and when you see that was closed with a message essentially stating, "Sorry, too vauge, try again", that is a good way to frustrate and alienate users.

I am a big supporter of Docker, and advocate many different types of projects to use it. I think that it would greatly help the project if issues like this were handled better. Specifically, I think when @tiborvass closed this issue, it should have been locked (so the "resolution" of the issue didn't get burried in the middle of the page), and included a reference to other issue(s) where the problem(s) could be discussed in the "more focused" fashion he was advocating for.

**Comment by foxx** [comment]
> to a particular problem that is an implementation detail of the current storage subsystem

@cpuguy83 The entire implementation of Docker is fundamentally flawed, and this issue is just one of many such issues. So unless you are planning on rewriting the entire Docker platform from scratch, then flattening images is the best you're going to get.

> The problem with this issue is that it provides a solution to a problem that yet has to be defined

@tiborvass I think it's pretty clear what the problem is, don't you?

**Comment by monokrome** [comment]
Problem: We have `n` layers when the actions performed for each one only need to be in 1 layer.
Solution: ?!?!?!?!?

**Comment by ohjames** [comment]
> The problem with this issue is that it provides a solution to a problem that yet has to be defined

I've seen answers like this every time someone proposes a feature like this. Clearly it is wrong. The translation is:

"LAH LAH LAH LAH LAH LAH LAH I CAN'T HEEEEEEEEEEEEEEEEEEEEAR YOUUUUUUUUUUU"

**Comment by ohjames** [comment]
> when you see that was closed with a message essentially stating, "Sorry, too vauge, try again", that is a good way to frustrate and alienate users.

I see hundreds of people defining a very very clear problem... Hundreds of users all in unanimous agreement that docker handles layers in a way that doesn't make sense to them. Yet the people on the inside actually developing it are the only ones who feel that hundreds of community members all agreeing with each other and stating the same thing haven't "defined" themselves.

Even if I did agree that the problem wasn't clearly well defined (and I definitely don't) the way the core developers have responded to the community basically shows contempt. As for the solutions, how `docker-squash` manages to be so slow and delay our build time for so long on such a tiny set of layers, I don't know... can't wait for rkt.

**Comment by zerthimon** [comment]
@ohjames +1
I feel the same thing. I asked for a few features before, and they all were rejected with the following reasons:
1. It will hurt portability
2. it will hurt security

When will this project realize, users don't like to be FORCED to have portability and security at the price of productivity.
How about adding the feature users ask for, so USER HAS THE CHOICE and DECIDES FOR HIMSELF if he wants to use it even if it hurts protability and security.

Can't wait for someone fork this project and make it more friendly to the users.

**Comment by justincormack** [comment]
There is an open PR for flattening https://github.com/docker/docker/pull/22641

**Comment by vdemeester** [comment]
*It took me a while to decide to answer something here, but I feel this issue also need some constructive comments*.

First, as @justincormack there is a PR for flattening (#22641) — thus maybe we could reopen that issue as we are trying to, *maybe*, have this built-in.

> How about adding the feature users ask for, so USER HAS THE CHOICE and DECIDES FOR HIMSELF if he wants to use it even if it hurts protability and security.

I'm gonna quote Nathan Leclaire here (from [The Dockerfile is not the source of truth for your image](http://nathanleclaire.com/blog/2014/09/29/the-dockerfile-is-not-the-source-of-truth-for-your-image/)). `Dockerfile`s and `docker build` is only **one* way

https://www.youtube.com/watch?v=IyuyA8rSBAo

**Comment by vdemeester** [comment]
*It took me a while to decide to answer something here, but I feel I need to pin-point some stuff.*

First, as @justincormack there is a PR for flattening (#22641) — thus maybe we could reopen that issue as we are trying to, *maybe*, have this built-in.

> How about adding the feature users ask for, so USER HAS THE CHOICE and DECIDES FOR HIMSELF if he wants to use it even if it hurts protability and security.

I'm gonna quote Nathan Leclaire here (from [The Dockerfile is not the source of truth for your image](http://nathanleclaire.com/blog/2014/09/29/the-dockerfile-is-not-the-source-of-truth-for-your-image/)). 

> The Dockerfile is a tool for creating images, but it is not the only weapon in your arsenal. 

`Dockerfile`s and `docker build` is only **one** way to build Docker images. It is the *default*/*built-in* one, but **you definitely have the choice** to build your image with other tooling (and there is some : [packer](https://www.packer.io/docs/builders/docker.html), [rocker](https://github.com/rocker-org/rocker), [dockramp](https://github.com/jlhawn/dockramp), [s2i](https://access.redhat.com/documentation/en/openshift-enterprise/3.0/using-images/chapter-2-source-to-image-s2i)… to only list a few). One of the focus of `Dockerfile` is *portability* and thus this is one of the main concern when discussing feature on `Dockerfile` and `docker`.

If you don't care about portability or if the `Dockerfile` possibilities are too limited for your use cases, again, repeating myself, you are free to use other tooling to build images. Docker does not *force you* to use `Dockerfile`s to build image — it's *just* the default, built-in way to do it.

On the `Dockerfile` and image building, I **highly recommend** people to watch a talk from Gareth Rushgrove at DockerCon16 : [The Dockerfile Explosion and the need for higher level tools](https://www.youtube.com/watch?v=IyuyA8rSBAo).

**Comment by rvs** [comment]
@justincormack Justin, what I really would love to see is very similar to #22641 but with a full power of git rebase (especially git rebase --interactive). Do you think it is feasible?

**Comment by docbill** [comment]
It is important to distinguish what is needed per this request, and what is
desired.   What is desired is to refactor docker images into a git or git
like repository view, so all the cool features one does with git would be
possible in docker.   For example, I have docker image for plex.   It is
basically a fedora base, with a download of the plex build onto.   I have
the container set to autobuild on the docker hub.   So every time the
Fedora base changes, it rebuilds, even though most of the time the plex
download does not change.   What annoys me about that one, is when I pull
the update the layer for adding the plex download is treated as brand new,
even though byte per byte is identical to the original delta.   With a git
like repository that could be handled, making a docker pull a much much
more efficient operation.

That is the desired...

However, the ask is simply to have a standard way to flatten an image.   So
lets say as part of my build I downloaded the plex source installed the
developer dependencies, compiled it, and then deleted the everything except
the actual build.   All the tools for the build would still be layers in
the image even though they would be inaccessible from my final image.   It
is a huge waste...  If the container could be flattened to remove unneeded
layers it would be much much more efficient use of space and bandwidth.

That is the required...

Don't say you can't do the required, because the desired is too much
work...   Just hit the low hanging fruit first, and everyone will be much
happier.

**Comment by jdmarshall** [comment]
@vdemeester 

| The Dockerfile is not the source of truth for your image

I think  some of the people asking for features like one are more comfortable with the truth in this statement than some of the people with 'Docker member' after their names.  There are, for instance, a number of security related issues that have been closed-won't-fix with the reason that `docker build` should be repeatable.

**Comment by cpuguy83** [comment]
For those interested, we just merged `--squash` on `docker build`.
This will squash the final result of the build to it's parent image (ie. the `FROM`).
https://github.com/docker/docker/pull/22641

**Comment by richard-scott** [comment]
It's been over a year now, any plans to introduce the "--squash" option into the CE codebase other than via the "--experimental" way.

**Comment by shinsenter** [comment]
I still need this feature in 2024, so I. created a shell script to do squashing docker image layers, without losing other instructions: https://github.com/shinsenter/docker-squash

**Comment by jdmarshall** [comment]
I just want to be able to squash some layers, so that I don't have to do a dozen lines of && for steps where the intermediate values are unimportant or even unwanted.

**Comment by swarren** [comment]
Multi-stage builds might help with the situation in the original bug description. Do all the work on the relevant file in one image, then in the build steps for the final image, either `COPY --from` or RUN --mount-type=bind,from=other_image` for one-layer access to those files. Of course, you still end up with a COPY or RUN for each dependency image, but it's better than nothing:-)

**Comment by jdmarshall** [comment]
Part of the attraction of Docker is the fact that you don't have to track the disposition of every single file that is output by a step in the process. It just shows up in the unionfs wherever files get written, be they in one directory or shotgunned across the entire filesystem.

Once you pull that rug out from under people by saying, "well manually copy the files you want using multi-stage builds", you're into self-sabotage.

Don't use avoiding the features of your system as a workaround for a bug in your system. From my perspective that's killing the goose that laid the golden eggs.
